### PR TITLE
Adding UnknownTopicPartition Akka message (fixes #4)

### LINKED ISSUE #4: FetchData does not properly handle topic/partition leadership change
The `FetchData` processor does not correctly handle the `UnknownTopicOrPartitionCode` error condition when topic/partition leadership changes. Currently the `handleUnknownTopicOrPartitionCode()` method of the `QueueProcessor` is executed when the `FetchData` processor encounters an `UnknownTopicOrPartitionCode`; this then sends an invalid Akka message (`RefreshBrokersAndLeaders`) to the `KafkaDispatcher` in an attempt to request reloading the topic/partition metadata. 

A new message (`UnknownTopicPartition`) should be defined in `KafkaDispatcherMessage.scala` which will be handled by the `KafkaDispatcher` to send a `RefreshBrokersAndLeaders` message to the parent `Router` worker.

### PR DESCRIPTION
**Pull request by uzair7**
@denisgrenader @samdehouck

### CODE REVIEW

File: src/main/scala/com/box/castle/router/kafkadispatcher/KafkaDispatcher.scala [hunk 1]
```diff
@@ -99,6 +99,10 @@ class KafkaDispatcher(boxSimpleConsumerFactory: CastleSimpleConsumerFactory,
         log.error(s"Broker is unreachable: $broker due to: ${t.getMessage}", t)
         context.parent ! BrokerUnreachable(broker)
       }
+      case UnknownTopicPartition(topicAndPartition) => {
+        log.error(s"Queue processor encountered unknown topic partition for: ${topicAndPartition}")
```
Comment (denisgrenader) [inline src/main/scala/com/box/castle/router/kafkadispatcher/KafkaDispatcher.scala]: log.error(s"KafkaDispatcher encountered unknown topic partition for: ${topicAndPartition}")

File: src/main/scala/com/box/castle/router/kafkadispatcher/processors/QueueProcessor.scala [hunk 2]
```diff
@@ -7,7 +7,7 @@ import com.box.castle.consumer.offsetmetadatamanager.OffsetMetadataManagerErrorC
 import com.box.castle.metrics.MetricsLogger
 import com.box.castle.router.exceptions.RouterFatalException
 import com.box.castle.router.kafkadispatcher.KafkaDispatcherRef
-import com.box.castle.router.kafkadispatcher.messages.{DispatchToKafka, KafkaBrokerUnreachable, KafkaResponse, UnexpectedFailure}
+import com.box.castle.router.kafkadispatcher.messages.{DispatchToKafka, KafkaBrokerUnreachable, KafkaResponse, UnexpectedFailure, UnknownTopicPartition}
 import com.box.castle.router.messages.{RefreshBrokersAndLeaders, RouterResult}
```
Comment (denisgrenader) [inline src/main/scala/com/box/castle/router/kafkadispatcher/processors/QueueProcessor.scala]: Remove RefreshBrokersAndLeaders from being imported

File: src/main/scala/com/box/castle/router/kafkadispatcher/messages/KafkaDispatcherMessage.scala [hunk 3]
```diff
@@ -79,3 +79,5 @@ private[kafkadispatcher] case class FetchOffsetKafkaResponse(response: OffsetRes
 
 private[kafkadispatcher] case class InternalTopicMetadataResponse(response: TopicMetadataResponse) extends KafkaResponse
 
+private[kafkadispatcher] case class UnknownTopicPartition[T <: DispatchToKafka](topicAndPartition: TopicAndPartition) extends KafkaDispatcherCommonMessage
```
Comment (denisgrenader) [inline src/main/scala/com/box/castle/router/kafkadispatcher/messages/KafkaDispatcherMessage.scala]: This sould just be

`private[kafkadispatcher] case class UnknownTopicPartition(topicAndPartition: TopicAndPartition) extends KafkaDispatcherCommonMessage`

**Comment by denisgrenader** [comment]
Can you please add a test that checks whether the Router gets a RefreshBrokersAndLeaders message when the queueprocessor encounters this problem?

**Review by denisgrenader — CHANGES_REQUESTED** [review]

File: src/main/scala/com/box/castle/router/kafkadispatcher/KafkaDispatcher.scala (hunk 1 above)
Comment (uzair7) [reply src/main/scala/com/box/castle/router/kafkadispatcher/KafkaDispatcher.scala]: Done

File: src/main/scala/com/box/castle/router/kafkadispatcher/processors/QueueProcessor.scala (hunk 2 above)
Comment (uzair7) [reply src/main/scala/com/box/castle/router/kafkadispatcher/processors/QueueProcessor.scala]: Done

File: src/main/scala/com/box/castle/router/kafkadispatcher/messages/KafkaDispatcherMessage.scala (hunk 3 above)
Comment (uzair7) [reply src/main/scala/com/box/castle/router/kafkadispatcher/messages/KafkaDispatcherMessage.scala]: Done

**Review by denisgrenader — APPROVED** [review]

**Comment by denisgrenader** [comment]
Looks good!

**Pull request merged**